### PR TITLE
Introduce `PlayerClientBrandEvent`

### PR DESCRIPTION
**Pull request by drunderscore**
Fired when the player sends the `minecraft:brand` (or equivalent) plugin message.  If the result is
`DENIED`, the player brand is not set (`setClientBrand` is not called),
and it is not forwarded to the backend.

### CODE REVIEW

**Comment by drunderscore** [comment]
Fix the wrong year in the copyright header

File: api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.google.common.base.Preconditions;
+import com.velocitypowered.api.event.ResultedEvent;
+import com.velocitypowered.api.event.connection.PluginMessageEvent;
+import com.velocitypowered.api.proxy.Player;
+
+/**
+ * Fired when a {@link Player} sends the <code>minecraft:brand</code> plugin message.
+ */
+public class PlayerClientBrandEvent implements ResultedEvent<PluginMessageEvent.ForwardResult> {
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java]: I don't see any purpose to allow this event to be cancelled. Why would you want to do that?

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java [hunk 2]
```diff
@@ -229,9 +230,14 @@ public boolean handle(PluginMessage packet) {
         player.getKnownChannels().removeAll(PluginMessageUtil.getChannels(packet));
         backendConn.write(packet.retain());
       } else if (PluginMessageUtil.isMcBrand(packet)) {
-        player.setClientBrand(PluginMessageUtil.readBrandMessage(packet.content()));
-        backendConn.write(PluginMessageUtil
-            .rewriteMinecraftBrand(packet, server.getVersion(), player.getProtocolVersion()));
+        var brand = PluginMessageUtil.readBrandMessage(packet.content());
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: Inconsistent with the rest of the code here, which does not use `var`.

**Review by astei — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java (hunk 1 above)
Comment (drunderscore) [reply api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java]: No clue! Honestly I just copied what most events did which was that they were cancelable. Can't think of a reason you'd want to cancel it :^)

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java (hunk 2 above)
Comment (drunderscore) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: no `var` :(
No problem, consistency over everything for sure.

**Comment by drunderscore** [comment]
Event is no longer cancelable, and the commit is now consistent with the reset of the codebase's lack of `var`.

Only issue is that the copyright year in `PlayerClientBrandEvent.java` is wrong (2018). However the licenser doesn't accept any other year, as the year is baked into `HEADER.txt` as 2018.

File: api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package com.velocitypowered.api.event.player;
+
+import com.velocitypowered.api.proxy.Player;
+
+/**
+ * Fired when a {@link Player} sends the <code>minecraft:brand</code> plugin message.
+ */
+public class PlayerClientBrandEvent {
```
Comment (A248) [inline api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java]: In tandem with preconditions (see other comment), a final class prevents sub-classes from overriding methods and breaking contracts. Other events do this as well.
```suggestion
public final class PlayerClientBrandEvent {
```

File: api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Creates a new instance.
+   *
+   * @param player the {@link Player} of the sent client brand
+   * @param brand the sent client brand
+   */
+  public PlayerClientBrandEvent(Player player, String brand) {
+    this.player = player;
+    this.brand = brand;
```
Comment (A248) [inline api/src/main/java/com/velocitypowered/api/event/player/PlayerClientBrandEvent.java]: I believe there should be preconditions here, i.e. Preconditions.checkNotNull as the other events in the API use.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java [hunk 5]
```diff
@@ -229,9 +230,8 @@ public boolean handle(PluginMessage packet) {
         player.getKnownChannels().removeAll(PluginMessageUtil.getChannels(packet));
         backendConn.write(packet.retain());
       } else if (PluginMessageUtil.isMcBrand(packet)) {
-        player.setClientBrand(PluginMessageUtil.readBrandMessage(packet.content()));
-        backendConn.write(PluginMessageUtil
-            .rewriteMinecraftBrand(packet, server.getVersion(), player.getProtocolVersion()));
+        String brand = PluginMessageUtil.readBrandMessage(packet.content());
+        server.getEventManager().fire(new PlayerClientBrandEvent(player, brand));
```
Comment (A248) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: After the event is fired, player.setClientBrand and backendConn.write should still be called.

Comment (drunderscore) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: whoops those accidentally disappeared in me making it non-cancelable!

**Comment by drunderscore** [comment]
Changes made based on @A248 feedback, commit message updated to not reference the cancelability of `PlayerClientBrandEvent` as it does not exist.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java [hunk 6]
```diff
@@ -229,9 +230,11 @@ public boolean handle(PluginMessage packet) {
         player.getKnownChannels().removeAll(PluginMessageUtil.getChannels(packet));
         backendConn.write(packet.retain());
       } else if (PluginMessageUtil.isMcBrand(packet)) {
+        String brand = PluginMessageUtil.readBrandMessage(packet.content());
+        server.getEventManager().fire(new PlayerClientBrandEvent(player, brand));
         player.setClientBrand(PluginMessageUtil.readBrandMessage(packet.content()));
```
Comment (A248) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]:
```suggestion
        player.setClientBrand(brand);
```
Strings are immutable.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java [hunk 7]
```diff
@@ -229,9 +230,11 @@ public boolean handle(PluginMessage packet) {
...
         backendConn.write(PluginMessageUtil
-            .rewriteMinecraftBrand(packet, server.getVersion(), player.getProtocolVersion()));
+                .rewriteMinecraftBrand(packet, server.getVersion(), player.getProtocolVersion()));
```
Comment (A248) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: Unnecessary formatting change

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java [hunk 8]
```diff
@@ -229,7 +230,9 @@ public boolean handle(PluginMessage packet) {
         player.getKnownChannels().removeAll(PluginMessageUtil.getChannels(packet));
         backendConn.write(packet.retain());
       } else if (PluginMessageUtil.isMcBrand(packet)) {
-        player.setClientBrand(PluginMessageUtil.readBrandMessage(packet.content()));
+        String brand = PluginMessageUtil.readBrandMessage(packet.content());
+        server.getEventManager().fire(new PlayerClientBrandEvent(player, brand));
```
Comment (A248) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]:
```suggestion
        server.getEventManager().fireAndForget(new PlayerClientBrandEvent(player, brand));
```
I forgot about this, it seems.

**Review by A248 — APPROVED** [review]
Besides fire versus fireAndForget, I approve. (disclaimer: I am not a core Velocity contributor)

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java (hunk 8 above)
Comment (drunderscore) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java]: Makes sense to me -- it's done above for `PlayerChannelRegisterEvent` as well.

**Review by astei — APPROVED** [review]
